text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+to run existing code but to make a specific file's API items available for use
+to do interactive whitebox testing of private code within a larger project.
+
+## Embedded or adjacent Lockfile
+
+[Lockfiles](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html)
+record the exact version used for every possible dependency to ensure
+reproducibility.  In particular, this protects against upgrading to broken
+versions and allows continued use of a yanked version.  As this time, the
+recommendation is for
+[`bin`s to persist their lockfile while `lib`s do not](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries).
+
+With multi-file packages, `cargo` writes a `Cargo.lock` file to the package
+directory.  As there is no package directory for single-file packages, we need
+to decide how to handle locking dependencies.
+
+Considerations
+- Sharing of single-file projects should be easy
+  - In "copy/paste" scenarios, like reproduction cases in issues, how often
+    have lockfiles been pertinent for reproduction?
+- There is an expectation of a reproducible Rust experience
+- Dropping of additional files might be frustrating for users to deal with (in
+  addition to making it harder to share it all)
+- We would need a way to store the lockfile for `stdin` without conflicting
+  with parallel runs
+- `cargo` already makes persisting of `Cargo.lock` optional for multi-file
+  packages, encouraging not persisting it in some cases
+- Newer users should feel comfortable reading and writing single-file packages
+- A future possibility is allowing single-file packages to belong to a
+  workspace at which point they would use the workspace's `Cargo.lock` file.
+  This limits the scope of the conversation and allows an alternative to
+  whatever is decided here.
+- Read-only single-file packages (e.g. running `/usr/bin/package.rs` without root privileges)
+
+> Disposition: Deferred.  We feel this can be handled later, either by checking
+> for a manifest field, like `workspace.lock`, or by checking if the lock
+> content exists (wherever it is stored).  The main constraint is that if we
+> want to embed the lock content in the `.rs` file, we leave syntactic room for
+> it.
+
+**Location 1: In `CARGO_TARGET_DIR`**
+
+The path would include a hash of the manifest to avoid conflicts.
+
+- Transient location, lost with a `cargo clean --manifest-path foo.rs`
+- Hard to find for sharing on issues, if needed
+
+**Location 2: In `$CARGO_HOME`**
+
+The path would include a hash of the manifest to avoid conflicts.
+
+- Transient location though not lost with `cargo clean --manifest-path foo.rs`
+- No garbage collection to help with temporary source files, especially `stdin`
+
+**Location 3: As `<file-stem>.lock`**
+
+Next to `<file-stem>.rs`, we drop a `<file-stem>.lock` file.   We could add a
+`_` or `.` prefix to distinguish this from the regular files in the directory.
```

Comment:
Why drop the extension? `script.rs.lock` seems orders of magnitue less likely to conflict with regular files, shows up *after* `script.rs` in file listings and has a more obvious connection to `script.rs`.

Reply:
As this is in the future possibilities section, the point is more to illustrate an idea.  We don't need to design this out at this time.